ugs-platform/ugs-platform-ugscore/src/main/java/com/willwinder/ugs/nbp/core/panels/FileBrowserPanel.java

```diff
@@ -0,0 +1,342 @@
+    }
+
+
+    private void changeDirectory(String path) {
+        File newDirectory = new File(path);
+        if (newDirectory.exists() && newDirectory.isDirectory()) {
+            setDirectory(newDirectory);
+        } else {
+            JOptionPane.showMessageDialog(this, "Directory does not exist: " + path, "Error", JOptionPane.ERROR_MESSAGE);
+        }
+    }
+
+    private void createChildren(DefaultMutableTreeNode node, File file, boolean showHidden) {
+        if (file == null) {
+            return;
+        }
+
+        File[] files = file.listFiles();
+        if (files != null) {
+            for (File child : files) {
+                if (!showHidden && (child.isHidden() || child.getName().startsWith(".") || child.getName().startsWith("$"))) {
+                    continue;
+                }
+                DefaultMutableTreeNode childNode = new DefaultMutableTreeNode(new FileNode(child));
+                node.add(childNode);
+                if (child.isDirectory()) {
+                    childNode.add(new DefaultMutableTreeNode(new FileNode(null, "Loading...")));
+                }
+            }
+        }
+    }
+
+    private void setPanelEnabled(boolean enabled) {
+        this.setEnabled(enabled);
+        goButton.setEnabled(enabled);
+        fileTree.setEnabled(enabled);
+        currentPathField.setEnabled(enabled);
+        showHiddenCheckBox.setEnabled(enabled);
+    }
+
+    private void setEnabledFromStatus(ControllerState controllerStateEvent) {
+        switch (controllerStateEvent) {
+            case DISCONNECTED, IDLE, UNKNOWN:
+                setPanelEnabled(true);
+                break;
+            default:
+                setPanelEnabled(false);
+                break;
+        }
+    }
+
+    @Override
+    public void UGSEvent(UGSEvent evt) {
+        if (evt instanceof ControllerStateEvent controllerStateEvent) {
+            setEnabledFromStatus(controllerStateEvent.getState());
+        } else if (evt instanceof ControllerStatusEvent controllerStatusEvent) {
+            setEnabledFromStatus(controllerStatusEvent.getStatus().getState());
+        }
+    }
+
+    private static class FileNode {
+        private final File file;
+        private final String displayName;
+
+        public FileNode(File file) {
+            this.file = file;
+            this.displayName = file.getName().isEmpty() ? file.getPath() : file.getName();
+        }
+
+        public FileNode(File file, String displayName) {
+            this.file = file;
+            this.displayName = displayName;
+        }
+
+        public boolean isDirectory() {
+            return file.isDirectory();
+        }
+
+        public File getFile() {
+            return file;
+        }
+
+        @Override
+        public String toString() {
+            return displayName;
+        }
+    }
+
+    public static class FileTreeCellRenderer extends DefaultTreeCellRenderer {
```

Comment:
Because I'm adding the `.` and `..` link paths, I wanted those to be viewed as folders. I scrapped some of the icons customizability and kept only the GCode file icons and let the OS/Java lib dictate the rest (but still forcing folder icons on `.` and `..`)